client/gutenberg/editor/controller.js

```diff
@@ -152,6 +156,12 @@ export const post = async ( context, next ) => {
 		const siteId = getSelectedSiteId( state );
 		const userId = getCurrentUserId( state );
 
+		// When copying a post, first invalidate the cache containing the `duplicatePostId` post,
+		// so that following `requestSitePost` will hit a freshly cached version of the post.
+		if ( !! duplicatePostId ) {
+			requestSitePost( siteId, duplicatePostId, postType, 0 );
```

Comment:
I understand this logic thanks to the great comment and because I have the context of the PR, but it seems to me a bit weird to request a post which we're not using at all, just because this will clear the cache for the next request. 

Why don't we request the non-cached version of the post directly here?

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/gutenberg/editor/main.jsx#L207

We can add a new `freshness` parameter to `getPost` that can be sent to `requestSitePost`:

```js
const getPost = ( siteId, postId, postType, freshness ) => {
	if ( siteId && postId && postType ) {
		const requestSitePostData = requestSitePost( siteId, postId, postType, freshness );
		return get( requestSitePostData, 'data', null );
	}

	return null;
};
```

So we can tell `getPost` we want the fresh post data:

```js
const postCopy = getPost( siteId, duplicatePostId, postType, 0 );
```

Reply:
That was my first thought as well, but unfortunately, that `getPost` method is called inside `mapStateToProps`, which means that it's called _a lot_ of times.

It's fine enough to do so if we hit the cache, as it's pretty much the same of a Redux selector.
But if we force the freshness to 0, then we would actually fire an actual network request every time the state changes, which ultimately results in crashing the browser.

I totally understand when you say that it's a confusing logic, but at the same time, it's very similar to our own Query Component pattern, where we force an API call somewhere (usually in the parent component), and use the data somewhere else (usually in the children).

I also tried to only call the 0-freshness `getPost` once in the Main, but introducing an internal state for it sounds more painful (and would waste more re-renders) than the proposed solution.

Reply:
> But if we force the freshness to 0, then we would actually fire an actual network request every time the state changes, which ultimately results in crashing the browser.

Indeed we would make a request, but only if we're duplicating a post, which I guess is what we want. `getPost` doesn't call to `requestSitePost` if `duplicatePostId` is not truthy.

Or am I missing something? Does the state still change a lot when `duplicatePostId` has a value?

Reply:
When we copy a post, `duplicatePostId` is passed as query param, and so it's always truthy from when the page starts loading until when the draft is saved and the URL is automatically updated.
Though, that's unrelated to how many times the `mapStateToProps`, and consequently `requestSitePost`, is called when the editor load.
`mapStateToProps` is called every time the Redux state changes, regardless of the `ownProps` argument values.

I've just tried just to be sure.
On copy post, the Main's `mapStateToProps` is called 13 times when we enter the editor, 5 times when we leave it, and 1 every time we do some header actions (e.g. save, publish, preview, etc.).

In other words, we would end up calling the posts API 13 times in roughly a second, and several other - completely unnecessary - times while we work on the post.

Reply:
Thanks for confirming!